Unable to Build NALU-CFD with Trilinos Configured for OpenMP

Hi NALU-CFD Team, been trying to get NALU-CFD running on one of our new test beds with OpenMP enabled for Trilinos. NALU-CFD will not compile because there is a forced use of Serial host-space which gives the errors below. It would be super-nice to be able to run Trilinos with OpenMP enabled and NALU-CFD in Serial if that's what is required, is that possible at all?

```
[  1%] Building CXX object CMakeFiles/nalu.dir/src/AssembleElemSolverAlgorithm.C.o
In file included from /home/sdhammo/nalu/trilinos/master-20170618/include/Kokkos_Core.hpp:53:0,
                 from /home/sdhammo/nalu/Nalu-master/include/KokkosInterface.h:12,
                 from /home/sdhammo/nalu/Nalu-master/include/SolverAlgorithm.h:13,
                 from /home/sdhammo/nalu/Nalu-master/include/AssembleElemSolverAlgorithm.h:12,
                 from /home/sdhammo/nalu/Nalu-master/src/AssembleElemSolverAlgorithm.C:10:
/home/sdhammo/nalu/trilinos/master-20170618/include/Kokkos_Serial.hpp: In instantiation of ‘typename std::enable_if<std::is_same<TagType, void>::value>::type Kokkos::Impl::ParallelFor<FunctorType, Kokkos::TeamPolicy<Properties ...>, Kokkos::Serial>::exec(Kokkos::Impl::HostThreadTeamData&) const [with TagType = void; FunctorType = sierra::nalu::AssembleElemSolverAlgorithm::execute()::<lambda(const TeamHandleType&)>; Properties = {Kokkos::Serial}; typename std::enable_if<std::is_same<TagType, void>::value>::type = void]’:
/home/sdhammo/nalu/trilinos/master-20170618/include/Kokkos_Serial.hpp:672:7:   required from ‘void Kokkos::Impl::ParallelFor<FunctorType, Kokkos::TeamPolicy<Properties ...>, Kokkos::Serial>::execute() const [with FunctorType = sierra::nalu::AssembleElemSolverAlgorithm::execute()::<lambda(const TeamHandleType&)>; Properties = {Kokkos::Serial}]’
/home/sdhammo/nalu/trilinos/master-20170618/include/Kokkos_Parallel.hpp:190:4:   required from ‘void Kokkos::parallel_for(const ExecPolicy&, const FunctorType&, const string&, typename Kokkos::Impl::enable_if<(! Kokkos::Impl::is_integral<ExecPolicy>::value)>::type*) [with ExecPolicy = Kokkos::TeamPolicy<Kokkos::Serial>; FunctorType = sierra::nalu::AssembleElemSolverAlgorithm::execute()::<lambda(const TeamHandleType&)>; std::__cxx11::string = std::__cxx11::basic_string<char>; typename Kokkos::Impl::enable_if<(! Kokkos::Impl::is_integral<ExecPolicy>::value)>::type = void]’
/home/sdhammo/nalu/Nalu-master/src/AssembleElemSolverAlgorithm.C:140:4:   required from here
/home/sdhammo/nalu/trilinos/master-20170618/include/Kokkos_Serial.hpp:640:9: error: no match for call to ‘(const sierra::nalu::AssembleElemSolverAlgorithm::execute()::<lambda(const TeamHandleType&)>) (Kokkos::Impl::ParallelFor<sierra::nalu::AssembleElemSolverAlgorithm::execute()::<lambda(const TeamHandleType&)>, Kokkos::TeamPolicy<Kokkos::Serial>, Kokkos::Serial>::Member)’
         m_functor( Member(data,ileague,m_league) );
         ^~~~~~~~~
/home/sdhammo/nalu/Nalu-master/src/AssembleElemSolverAlgorithm.C:99:79: note: candidate: sierra::nalu::AssembleElemSolverAlgorithm::execute()::<lambda(const TeamHandleType&)>
   Kokkos::parallel_for(team_exec, [&](const sierra::nalu::TeamHandleType& team)
                                                                               ^
/home/sdhammo/nalu/Nalu-master/src/AssembleElemSolverAlgorithm.C:99:79: note:   no known conversion for argument 1 from ‘Kokkos::Impl::ParallelFor<sierra::nalu::AssembleElemSolverAlgorithm::execute()::<lambda(const TeamHandleType&)>, Kokkos::TeamPolicy<Kokkos::Serial>, Kokkos::Serial>::Member {aka Kokkos::Impl::HostThreadTeamMember<Kokkos::Serial>}’ to ‘const TeamHandleType& {aka const Kokkos::Impl::HostThreadTeamMember<Kokkos::OpenMP>&}’
```

--- a/include/KokkosInterface.h
+++ b/include/KokkosInterface.h
@@ -15,15 +15,17 @@ namespace sierra {
 namespace nalu {
 
 using HostSpace = Kokkos::HostSpace;
-using DeviceSpace = Kokkos::DefaultExecutionSpace;
+//!!!! Important: the 'Kokkos::Serial' here makes sure we only run serial. !!!!
+// using DeviceSpace = Kokkos::DefaultExecutionSpace;
+using DeviceSpace = Kokkos::Serial;
 using DeviceShmem = DeviceSpace::scratch_memory_space;
-typedef Kokkos::Schedule<Kokkos::Dynamic> DynamicScheduleType;
-typedef typename Kokkos::TeamPolicy<typename Kokkos::DefaultExecutionSpace, DynamicScheduleType>::member_type TeamHandleType;
+using DynamicScheduleType = Kokkos::Schedule<Kokkos::Dynamic>;
+using TeamHandleType = Kokkos::TeamPolicy<DeviceSpace, DynamicScheduleType>::member_type;
+
 template <typename T>
 using SharedMemView = Kokkos::View<T, Kokkos::LayoutRight, DeviceShmem, Kokkos::MemoryUnmanaged>;
 
-//!!!! Important: the 'Kokkos::Serial' here makes sure we only run serial. !!!!
-using DeviceTeamPolicy = Kokkos::TeamPolicy<Kokkos::Serial>;
+using DeviceTeamPolicy = Kokkos::TeamPolicy<DeviceSpace>;
 using DeviceTeam = DeviceTeamPolicy::member_type;
 
 inline DeviceTeamPolicy get_team_policy(const size_t sz, const size_t bytes_per_team,

--- a/include/KokkosInterface.h
+++ b/include/KokkosInterface.h
@@ -17,7 +17,12 @@ namespace nalu {
 using HostSpace = Kokkos::HostSpace;
 //!!!! Important: the 'Kokkos::Serial' here makes sure we only run serial. !!!!
 // using DeviceSpace = Kokkos::DefaultExecutionSpace;
+#ifdef KOKKOS_HAVE_SERIAL
 using DeviceSpace = Kokkos::Serial;
+#else
+#error "Kokkos does not have Serial space enabled"
+#endif
+
 using DeviceShmem = DeviceSpace::scratch_memory_space;
 using DynamicScheduleType = Kokkos::Schedule<Kokkos::Dynamic>;
 using TeamHandleType = Kokkos::TeamPolicy<DeviceSpace, DynamicScheduleType>::member_type;

--- a/unit_tests/UnitTestKokkosUtils.h
+++ b/unit_tests/UnitTestKokkosUtils.h
@@ -43,7 +43,7 @@ void kokkos_bucket_loop(const stk::mesh::BucketVector& buckets, LOOP_BODY inner_
 template<class LOOP_BODY>
 void kokkos_thread_team_bucket_loop(const stk::mesh::BucketVector& buckets, LOOP_BODY inner_loop_body)
 {
-    Kokkos::parallel_for(Kokkos::TeamPolicy<Kokkos::DefaultExecutionSpace>(buckets.size(), Kokkos::AUTO), KOKKOS_LAMBDA(const sierra::nalu::TeamHandleType& team)
+    Kokkos::parallel_for(sierra::nalu::DeviceTeamPolicy(buckets.size(), Kokkos::AUTO), KOKKOS_LAMBDA(const sierra::nalu::TeamHandleType& team)
     {
         const stk::mesh::Bucket& bkt = *buckets[team.league_rank()];
         Kokkos::parallel_for(Kokkos::TeamThreadRange(team, bkt.size()), [&](const size_t& j)
@@ -57,7 +57,7 @@ template<class LOOP_BODY>
 void kokkos_thread_team_bucket_loop_with_topo(const stk::mesh::BucketVector& buckets,
                                     const LOOP_BODY& inner_loop_body)
 {
-    Kokkos::parallel_for(Kokkos::TeamPolicy<Kokkos::DefaultExecutionSpace>(buckets.size(), Kokkos::AUTO), KOKKOS_LAMBDA(const sierra::nalu::TeamHandleType& team)
+    Kokkos::parallel_for(sierra::nalu::DeviceTeamPolicy(buckets.size(), Kokkos::AUTO), KOKKOS_LAMBDA(const sierra::nalu::TeamHandleType& team)
     {
         const stk::mesh::Bucket& bkt = *buckets[team.league_rank()];
         stk::topology topo = bkt.topology();

